File: src/lib/scala/Slate.Core/src/main/scala/slate/core/mods/Module.scala
/**
  * <slate_header>
  * author: Kishore Reddy
  * url: https://github.com/kishorereddy/scala-slate
  * copyright: 2016 Kishore Reddy
  * license: https://github.com/kishorereddy/scala-slate/blob/master/LICENSE.md
  * desc: a scala micro-framework
  * usage: Please refer to license on github for more info.
  * </slate_header>
  */
package slate.core.mods

import slate.core.common.AppContext


abstract class Module( val appCtx:AppContext,
                       val modCtx:ModuleContext)
{
  val info:ModuleInfo

  /**
    * hook to initialize
    */
  def init():Unit = {
  }


  def register():Unit = {
  }


  /**
    * install this module
    */
  def install():Unit = {
    if(info.isDbDependent ){
      info.models.fold(Unit)( models => {
        models.foreach( modelName => {
          modCtx.modelService.install(modelName, info.version, "", "")
        })
        Unit
      })
    }
  }


  def script():List[String] = {
    if(info.isDbDependent ){
      info.models.fold(List[String]())( models => {
        models.map(modelName => {
          modCtx.modelService.generateSql(modelName, info.version)
          modelName
        })
      })
    }
    else
      List[String]()
  }


  def toItem(): Mod = {
    val item = new Mod(
      name          = info.name,
      desc          = info.desc,
      version       = info.version,
      isInstalled   = info.isInstalled,
      isEnabled     = info.isEnabled,
      isDbDependent = info.isDbDependent,
      totalModels   = info.totalModels,
      source        = info.source,
      dependencies  = info.dependencies
    )
    item
  }
}

File: src/lib/kotlin/slatekit-core/src/main/kotlin/slatekit/core/mods/Module.kt
/**
 * <slate_header>
 * url: www.slatekit.com
 * git: www.github.com/code-helix/slatekit
 * org: www.codehelix.co
 * author: Kishore Reddy
 * copyright: 2016 CodeHelix Solutions Inc.
 * license: refer to website and/or github
 * about: A tool-kit, utility library and server-backend
 * mantra: Simplicity above all else
 * </slate_header>
 */

package slatekit.core.mods

import slatekit.core.common.AppContext

abstract class Module(val appCtx: AppContext,
                      val modCtx: ModuleContext) {
    abstract val info: ModuleInfo

    /**
     * hook to initialize
     */
    open fun init(): Unit {}


    open fun register(): Unit {}


    /**
     * install this module
     */
    open fun install(): Unit {
        if (info.isDbDependent) {
            info.models?.let { models ->
                models.forEach { modelName ->
                    modCtx.setup.install(modelName, info.version, "", "")
                }
            }
        }
    }


    /**
     * script out all the sql for all the modules.
     */
    open fun script(): List<String> {
        val result = if (info.isDbDependent) {
            info.models?.let { models ->
                models.map { modelName ->
                    val result = modCtx.setup.generateSql(modelName, info.version)
                    result.msg ?: modelName
                }
            } ?: listOf<String>()
        }
        else
            listOf<String>()

        return result
    }


    fun toItem(): Mod {
        val item = Mod(
                name = info.name,
                desc = info.desc,
                version = info.version,
                isInstalled = info.isInstalled,
                isEnabled = info.isEnabled,
                isDbDependent = info.isDbDependent,
                totalModels = info.totalModels,
                source = info.source,
                dependencies = info.dependencies
        )
        return item
    }
}